plumbing/http/http.go

```diff
@@ -0,0 +1,327 @@
+			&transport.ReceivePackOptions{
+				GitProtocol:   version,
+				AdvertiseRefs: false,
+				StatelessRPC:  true,
+			})
+	}
+	if err != nil {
+		logf(errorLog, "error processing request: %v", err)
+		renderStatusError(w, http.StatusInternalServerError)
+		return
+	}
+
+	flusher, ok := w.(http.Flusher)
+	if !ok {
+		logf(errorLog, "http writer must implement http.Flusher interface")
+		// The writer must implement the Flusher interface which the default
+		// http.ResponseWriter does.
+		renderStatusError(w, http.StatusBadRequest)
+		return
+	}
+
+	// Write the response to the client in chunks.
+	p := make([]byte, 1024)
+	for {
+		nr, err := out.Read(p)
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		nw, err := w.Write(p[:nr])
+		if err != nil {
+			logf(errorLog, "error writing response: %v", err)
+			renderStatusError(w, http.StatusInternalServerError)
+			return
+		}
+		if nr != nw {
+			logf(errorLog, "error writing response: %v", err)
+			renderStatusError(w, http.StatusInternalServerError)
+			return
+		}
+		flusher.Flush()
+	}
+}
+
+func sendFile(w http.ResponseWriter, r *http.Request, contentType string) {
+	ctx := r.Context()
+	st := ctx.Value(contextKey("storer")).(storage.Storer)
+	fss, ok := st.(storer.FilesystemStorer)
+	if !ok {
+		renderStatusError(w, http.StatusNotFound)
+		return
+	}
+
+	file := ctx.Value(contextKey("file")).(string)
+	fs := fss.Filesystem()
+	f, err := fs.Open(file)
+	if err != nil {
+		renderStatusError(w, http.StatusNotFound)
+		return
+	}
+
+	defer f.Close() //nolint:errcheck
+	stat, err := fs.Stat(file)
+	if err != nil {
+		renderStatusError(w, http.StatusNotFound)
+		return
+	}
+
+	w.Header().Set("Content-Type", contentType)
+	w.Header().Set("Content-Length", fmt.Sprintf("%d", stat.Size()))
+	w.Header().Set("Last-Modified", stat.ModTime().Format(http.TimeFormat))
+	io.Copy(w, f) //nolint:errcheck
+}
+
+func getTextFile(w http.ResponseWriter, r *http.Request) {
+	hdrNocache(w)
+	sendFile(w, r, "text/plain; charset=utf-8")
+}
+
+func getInfoRefs(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	st := ctx.Value(contextKey("storer")).(storage.Storer)
+	errorLog := ctx.Value(contextKey("errorLog")).(*log.Logger)
+	service := transport.Service(r.FormValue("service"))
+	version := r.Header.Get("Git-Protocol")
+
+	if service != "" {
+		hdrNocache(w)
+		w.Header().Set("Content-Type", fmt.Sprintf("application/x-git-%s-advertisement", service.Name()))
+
+		var err error
+		switch service {
+		case transport.UploadPackService:
+			err = transport.UploadPack(ctx, st, nil, ioutil.WriteNopCloser(w),
+				&transport.UploadPackOptions{
+					GitProtocol:   version,
+					AdvertiseRefs: true,
+					StatelessRPC:  true,
+				},
+			)
+		case transport.ReceivePackService:
+			err = transport.ReceivePack(ctx, st, nil, ioutil.WriteNopCloser(w),
+				&transport.ReceivePackOptions{
+					GitProtocol:   version,
+					AdvertiseRefs: true,
+					StatelessRPC:  true,
+				},
+			)
```

Comment:
What if the service isn't either of those

Reply:
Good catch, we should return an error